Rollup support for TSVB

alexwizp opened the pull request:
## Summary
Roll-up functionality support for TSVB. 
#### Index pattern and rollups
- [x] In TSVB under Panel Options in the Index Pattern field, users will be able to add rollup index, regular index pattern or a mix or roll up and index pattern.
  - Same restrictions as Visual editor apply: TSVB will not support rollups names using wildcards; Only 1 rollup can be used 
- [x] Based on the input in Panel Options in the Index Pattern field, TSVB will be able to query:
  - regular indices
  - rollup index 
  - index combining a rollup and index pattern
- [x] Using TSVB user will be able to create multiple y-axis with different index patterns or rollups indices in the same chart.
- [x] Users specifying the index in TSVB will need to be explicit regarding the rollup name, TSVB index field will not include rollup index when using a wildcard
#### Aggregations
- [ ] Similar to kibana visual editor TSVB will limit aggregations available in the TSVB UI based on the capabilities of the roll-up index pattern using the rollup capabilities api
- [x] TSVB will limit fields in rollup based on aggregation type
- [x] Pipeline aggregations (derivatives, moving average, etc) should include rolled up data (TSVB is often used to create an advanced calculation, aggregation supported should include the pipeline aggregations)
- [x] TSVB unique aggregations should work on rolled up data. For example:
  - Math
  - Serial aggregation
- [ ] All TSVB visualizations should work with rollups
- [x] TSVB group by will not support filters and will only support `term` or `everything`
#### Time intervals
- [x] TSVB intervals should be a multiple of what is configured in the roll-up job. Since intervals are typed into TSVB it will be good to warn user ahead of time what is the minimal interval and not allow smaller intervals
- [x] It will be possible to set >= interval using the minimal intervals, which will still work by multiplying the minimal intervals
- [x] TSVB chart should support zoom in to a narrow time range or to the most granular possible in case it is a combination chart https://github.com/elastic/elasticsearch/issues/35744
- [x] In case user zoom in to an interval smaller than the minimum interval in a roll-up only chart, TSVB should use the smallest interval by default
#### Time zones
- [x] A rollup job has always a configured time zone for the date field it rolled up the data on. The date_histogram on that field mus NOT use the Kibana configured timezone, but the timezone that is configured in the rollup job instead, since the request will fail otherwise.

### Checklist

sulemanof committed:
Added a feature of rollup search on the UI side

Signed-off-by: Alexey Antonov <alexwizp@gmail.com>

alexwizp committed:
Rollup Feature - initial commit

sulemanof committed:
Revert "Added a feature of rollup search on the UI side"

This reverts commit 9568b0970b16f5102f50b748bb4d691a8612c2c2.

# Conflicts:
#	src/legacy/core_plugins/metrics/public/components/index_pattern.js

alexwizp committed:
Remove the 'label' property from the search strategies

sulemanof committed:
Changed search by strategy from the last

alexwizp committed:
add single search request

alexwizp committed:
rollup_search_strategy add base implementation of isViable method

alexwizp committed:
rollup_search_strategy add base implementation of isViable method -fix

sulemanof committed:
Changed requests due to search request type

alexwizp committed:
refactoring of import Base classes / remove '../../../../../../

markov00 commented on x-pack/plugins/rollup/server/lib/search_strategies/rollup_search_strategy.js:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    async isViable(req, indexPattern) {
+      return await await this.isRollupJobExists(req, indexPattern);
```
if I’m not wrong isRollupJobExists returns a promise that resolves to true or false, so 1 await is enough

alexwizp committed:
remove extra await

markov00 commented on x-pack/plugins/rollup/server/lib/search_strategies/rollup_search_strategy.js:
```diff
@@ -0,0 +1,36 @@
+      }));
+
+      return Promise.all(requests)
+        .then((responses) => responses.some((response, index) => Boolean(response[indices[index]])))
```
I think that @jen-huang told us that to use the rollup search we need to have at maximum 1 rollup index on the indexpattern. It will not work if the user wrote an index-pattern string that includes 2 rollup indexes. @jen-huang correct me if I'm wrong

alexwizp replied:
From our requirements: 

> Based on the input in Panel Options in the Index Pattern field, TSVB will be able to query:
> - regular indices
> - rollup index 
> - index combining a rollup and index pattern

I need this one for add supports of:  index combining a rollup and index pattern

jen-huang replied:
@markov00 @alexwizp Correct, the rollup search api only supports a maximum of one rollup index, along with any number of regular indices. See the 'rules' section in docs: https://www.elastic.co/guide/en/elasticsearch/reference/master/rollup-search.html#_path_parameters_69

This method could be changed to `numberOfRollupJobs()`, and returning the number of rollup index matches. `isViable()` can then check if `numberOfRollupJobs() === 1`.

alexwizp replied:
Got it, will be fixed

markov00 reviewed (commented):
Overall code LGTM.
Please change all default exports to named exports as written in our [styleguide](https://github.com/elastic/kibana/blob/master/style_guides/js_style_guide.md#use-named-exports-only).

alexwizp commented on src/legacy/core_plugins/metrics/server/lib/vis_data/series/get_request_params.js:
```diff
@@ -16,20 +16,24 @@
-    index: indexPattern,
-    ignoreUnavailable: true,
+    ...buildRequestBody(req, panel, series),
+    timeout: '90s'
```
@timroes  @markov00  
Am I right that we should use shardTimeout here instead of hardcoded 90s

See the following link for more details:
https://github.com/elastic/kibana/issues/17576

markov00 replied:
Yes you are right. To have a clean history it's better to add this fix on a different PR

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7257/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7442/)

alexwizp committed:
rollup_search_strategy. Refactoring of isRollupJobExists method

alexwizp committed:
remove question

sulemanof committed:
Add support of annotations and table data

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7462/)

sulemanof committed:
Fix TSVB search requests should have a timeout

# Conflicts:
#	src/legacy/core_plugins/metrics/server/lib/vis_data/get_annotations.js
#	src/legacy/core_plugins/metrics/server/lib/vis_data/series/get_request_params.js

sulemanof committed:
Add unit test

alexwizp committed:
skeleton for adding Search Strategy restrictions

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7705/)

sulemanof committed:
Add rollup search capabilities

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7894/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7902/)

alexwizp committed:
apply search strategy for annotations request

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7920/)

alexwizp committed:
set fields capabilities for rollup strategy

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7955/)

alexwizp committed:
add timezone, interval into SearchCapabilities

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8136/)

sulemanof committed:
Add fields from capabilities

alexwizp committed:
add timezone, interval into SearchCapabilities

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8165/)

alexwizp committed:
fix default timezone

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8175/)

alexwizp committed:
Merging of two Rollup Jobs was removed

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8192/)

alexwizp committed:
move getFieldsForWildcard to searchStrategy

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8251/)

alexwizp committed:
apply getEsShardTimeout for annorations/get_request_params,  series/get_request_params

timroes commented on src/legacy/core_plugins/metrics/index.js:
```diff
@@ -21,14 +21,16 @@ import { resolve } from 'path';
 
 export default function (kibana) {
   return new kibana.Plugin({
+    id: 'metrics',
```
Could we rename that to `tsvb`, since it's the name we usually use for the feature.

timroes replied:
@spalger I am always confused what the `id` here is doing. Will that also influence the prefix for settings for that plugin, or will that prefix still be determined by the package name?

alexwizp replied:
@timroes  Please see 
>    x-pack\plugins\rollup\server\lib\search_strategies\register_rollup_search_strategy.js

It's just a ID of plugin which we add into server.plugins object

renamed to tsvb. Thanks

alexwizp replied:
Good question, @spalger do you know? Because it's not working without adding the 'id' here

sulemanof replied:
@alexwizp @timroes I've found out that the `id` doesn't influence on requiring plugins, since they are included by plugins directory names

alexwizp replied:
resolved

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8275/)

timroes commented on src/legacy/core_plugins/metrics/server/lib/search_strategies/search_strategies_register.js:
```diff
@@ -0,0 +1,60 @@
+
+const strategies = [];
+
+export default class SearchStrategiesRegister {
```
As far as I see among the other code, we're trying to access `getViableStrategy` as a static method. If that's really needed we don't need to create a class here at all, that's just weird OOP pattern. But I think we should rather try to get in the places we are needing that method, get it from the `server` object instead. I think the `server` object is actually accessible via the request (`req`) object, that we have in all those places.

alexwizp replied:
Let's make all methods static. I've updated the PR. Not sure that it's a good idea to add this one into the request object. It's not obvious.  @timroes  what do you think?

timroes commented on src/legacy/core_plugins/metrics/server/lib/search_strategies/searh_requests/search_request.js:
```diff
@@ -0,0 +1,37 @@
+export default class SearchRequest extends AbstractSearchRequest {
+  getSearchRequestType(options) {
+    const isMultiSearch = Array.isArray(options.body);
+    const SearchRequest = isMultiSearch ? MultiSearchRequest : SingleSearchRequest;
```
Do we really need to make that distinction? Couldn't we just use `msearch` also for the cases we get in a single body?

alexwizp replied:
Are you sure we need to fix it? 'search' works faster than 'msearch' for single queries.  But you are right we can use msearh for all cases. @timroes  should we fix that?

timroes commented on x-pack/plugins/rollup/index.js:
```diff
@@ -20,7 +21,7 @@ export function rollup(kibana) {
     id: PLUGIN.ID,
     publicDir: resolve(__dirname, 'public'),
-    require: ['kibana', 'elasticsearch', 'xpack_main'],
+    require: ['kibana', 'elasticsearch', 'xpack_main', 'metrics'],
```
I am a bit worried about that. @epixa does that mean, that the `rollup` plugin won't startup as long as any of the required plugins is disabled via the `kibana.yml`?

epixa replied:
I'm not 100% on the behavior for the legacy plugin `require`, but I'm pretty sure all of Kibana will fail to start up if there's a plugin enabled that `require`s a plugin that is disabled. We don't do any automatic disabling in the legacy world though.

timroes replied:
@alexwizp Could you please try out what happens if you put `metrics.enabled: false` into your `kibana.yml`, it should not include TSVB then, but will it still come up?

sulemanof replied:
Hi @timroes. In that case the server will crash:
![image](https://user-images.githubusercontent.com/31325372/51736303-c4c7ed80-209a-11e9-8578-a94c127fdb82.png)

alexwizp replied:
@timroes  Thanks! it's not working! We'll fix

alexwizp replied:
resolved

timroes commented on x-pack/plugins/rollup/server/lib/search_strategies/rollup_search_capabilities.js:
```diff
@@ -0,0 +1,40 @@
+
+    getDateHistogramAggregation() {
+      const dateHistogramFields = this.getFieldsByAggregationType('date_histogram');
+      const keys = Object.keys(dateHistogramFields);
```
Could you explain what that's supposed to do? This looks a bit like we're implicitly relying on the order that `Object.keys` are returning the keys, which is NOT guaranteed to be in any specific order.

alexwizp replied:
Agree with you, it's not obvious... I'll fix to make more readable. 

Let me explain: 
getFieldsByAggregationType method returns all fields which supports passed aggregation in the following format:

```
{
   field_name1: {
        agg: 'histogram',
        option1: 'option1', 
        option2: 'option2' ,
   },
    field_name2: {
        agg: 'histogram',
        option1: 'option1', 
        option2: 'option2',
   }
}
```
Rollup Job can have only one 'date_histogram'. So I take the first value from that object.

alexwizp replied:
Fixed

alexwizp committed:
rename metrics -> tsvb

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8297/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8445/)

alexwizp committed:
search_strategies_register refactoring: move 'add' method from class

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8468/)

sulemanof committed:
Add merge rollup capabilities with fields

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8471/)

alexwizp committed:
Add merge rollup capabilities with fields - small fixes

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8535/)

alexwizp commented on x-pack/plugins/rollup/server/routes/api/index_patterns.js:
```diff
@@ -66,62 +66,10 @@ export function registerFieldsForWildcardRoute(server) {
         // Keep meta fields
         metaFields.forEach(field => fieldsFromFieldCapsApi[field] && rollupFields.push(fieldsFromFieldCapsApi[field]));
 
-        // Merge rollup capabilities information with field information
```
@jen-huang  Could you please review that? Looks like we can reuse this code.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8752/)

alexwizp committed:
Add support of 'Everything' aggregation for Rollup Search

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8793/)

sulemanof committed:
Return back metrics plugin

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8806/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8968/)

alexwizp committed:
remove 'metrics' from the X-pack\rollup require

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9030/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9237/)

sulemanof committed:
Fix test cases

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9252/)

alexwizp committed:
fix broken test: fail: "apis InfraOps GraphQL Endpoints metrics should basically work"

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9259/)

alexwizp committed:
rollup search - split by terms is not working

sulemanof committed:
Add count metric

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9444/)

alexwizp committed:
/get_bucket_size.js. Add support of 'auto' interval, Add support of gte intervals  e.g.:  >=1m

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9454/)

alexwizp committed:
fix build_request_body test

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9461/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9674/)

alexwizp committed:
[Rollup] [Phase 1] Error handling - rollup search errors should be more user friendly

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9707/)

alexwizp committed:
[Rollup] [Phase 1] Table View - research the query to ES - sorting is not wokring

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9755/)

alexwizp committed:
Merge #26006 into rollup

# Conflicts:
#	src/legacy/core_plugins/metrics/server/lib/vis_data/annorations/build_request_body.js
#	src/legacy/core_plugins/metrics/server/lib/vis_data/get_annotations.js
#	src/legacy/core_plugins/metrics/server/lib/vis_data/get_series_data.js
#	src/legacy/core_plugins/metrics/server/lib/vis_data/get_table_data.js
#	src/legacy/core_plugins/metrics/server/lib/vis_data/request_processors/annotations/query.js
#	src/legacy/core_plugins/metrics/server/lib/vis_data/series/__tests__/build_request_body.js
#	src/legacy/core_plugins/metrics/server/lib/vis_data/series/build_request_body.js
#	src/legacy/core_plugins/metrics/server/lib/vis_data/series/get_request_params.js

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10027/)

sulemanof committed:
Add table view support

alexwizp committed:
fix broken build

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10069/)

alexwizp committed:
fix broken build

alexwizp committed:
Merge remote-tracking branch 'upstream/master' into rollup

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10272/)

alexwizp committed:
[Rollup] [Phase 1] - write new tests  (rollup_search_request, rollup_search_strategy)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10295/)

alexwizp committed:
[Rollup] [Phase 1] - write tests for rollup_search_capabilities

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10339/)

alexwizp commented:
retest

sulemanof committed:
Add test on default_search_capabilities, abstract_search_strategy, search_strategies_register

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10350/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10354/)

alexwizp committed:
Merge remote-tracking branch 'upstream/master' into rollup

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10613/)

sulemanof committed:
Add test cases for search_requests folder

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10627/)

alexwizp committed:
[Rollup] [Phase 1] - write tests for rollup_search_strategy

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10654/)

AlonaNadler commented:
I played with the PR, few comments:
* TSVB suggests fields which they are not valid for the function. Example, I selected average and then in the field it suggests all metrics fields even though not all can perform average. After selecting the function it should only suggest field which can perform the function based on the rollup configuration. 
* when I played with the intervals I quickly got to this error
![image](https://user-images.githubusercontent.com/32550316/52379949-d5744e00-2a20-11e9-809c-6e616a49d4e5.png)
It happened when I changed to 1d interval rollup was set an hour previously, I don't think it valid, even there is partial data from the day, it supposes to work
* Filter panel and group by filter didn't work
* When I change to group by a filter and return to everything the chart panel is empty it only refresh once I select something in the UI, anything
![image](https://user-images.githubusercontent.com/32550316/52380748-e2467100-2a23-11e9-8588-2cc06894fbe0.png)
* I configured a value count for a specific field but couldn't make it work with the value count aggregation
* Pipeline aggregation (derivatives, cumulative sum didn't work for me)
![image](https://user-images.githubusercontent.com/32550316/52380871-4a955280-2a24-11e9-82e0-0f8e1902b375.png)

Overall my preference is that whatever we don't support (aggregation, group by, fields for specific calculation not in the rollup configuration) we will disable otherwise it will create a frustrating experience where I need to work with the rollup configuration open to understanding what can I do

sulemanof committed:
Merge remote-tracking branch 'upstream/master' into rollup

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11501/)

sulemanof committed:
FIx broken build

timroes commented:
> TSVB suggests fields which they are not valid for the function. Example, I selected average and then in the field it suggests all metrics fields even though not all can perform average. After selecting the function it should only suggest field which can perform the function based on the rollup configuration.

As of previous discussions this is not in phase 1 and for now we'll just fail for those.

sulemanof commented:
Hi @AlonaNadler ! 
> * TSVB suggests fields which they are not valid for the function. Example, I selected average and then in the field it suggests all metrics fields even though not all can perform average. After selecting the function it should only suggest field which can perform the function based on the rollup configuration.

Will be implemented in a phase 2 as a UI changes

> * when I played with the intervals I quickly got to this error
>   ![image](https://user-images.githubusercontent.com/32550316/52379949-d5744e00-2a20-11e9-809c-6e616a49d4e5.png)
>   It happened when I changed to 1d interval rollup was set an hour previously, I don't think it valid, even there is partial data from the day, it supposes to work

The  problem here that the rollup search endpoint fails itself on such intervals, so we would ask a suggestion how to workaround with such failures

> * Filter panel and group by filter didn't work

As far as we investigate this filed, filters are not working on a rollup search.
These fields will be disabled in a phase 2 as a UI changes

> * When I change to group by a filter and return to everything the chart panel is empty it only refresh once I select something in the UI, anything
>   ![image](https://user-images.githubusercontent.com/32550316/52380748-e2467100-2a23-11e9-8588-2cc06894fbe0.png)

There was a regression on master branch, already fixed and merged.

> * I configured a value count for a specific field but couldn't make it work with the value count aggregation

Please, provide more details to dive in that

> * Pipeline aggregation (derivatives, cumulative sum didn't work for me)
>   ![image](https://user-images.githubusercontent.com/32550316/52380871-4a955280-2a24-11e9-82e0-0f8e1902b375.png)

We will work on it

> Overall my preference is that whatever we don't support (aggregation, group by, fields for specific calculation not in the rollup configuration) we will disable otherwise it will create a frustrating experience where I need to work with the rollup configuration open to understanding what can I do
> 
> I didnt get to check combination of rollup and regular index pattern, and for a index pattern that combines both

sulemanof commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11525/)

AlonaNadler commented:
> * when I played with the intervals I quickly got to this error
>   ![image](https://user-images.githubusercontent.com/32550316/52379949-d5744e00-2a20-11e9-809c-6e616a49d4e5.png)
>   It happened when I changed to 1d interval rollup was set an hour previously, I don't think it valid, even there is partial data from the day, it supposes to work

The problem here that the rollup search endpoint fails itself on such intervals, so we would ask a suggestion how to workaround with such failures
@jen-huang was it done like that on purpose? 

* Good **temporary** call with not disabling the UI if it adds a lot of complexity, it's important we will solve it in phase 2. Currently, it requires to work with another window open looking at the rollup configuration

* Regarding Value count, 
![image](https://user-images.githubusercontent.com/32550316/52431496-33567380-2abd-11e9-8717-7a9eed4a1bd1.png)
![image](https://user-images.githubusercontent.com/32550316/52431540-45381680-2abd-11e9-8628-16a35e525565.png)

jen-huang commented:
> It happened when I changed to 1d interval rollup was set an hour previously, I don't think it valid, even there is partial data from the day, it supposes to work

> The problem here that the rollup search endpoint fails itself on such intervals, so we would ask a suggestion how to workaround with such failures

@AlonaNadler @sulemanof 
If I'm reading the issue correctly, the query is failing for certain `date_histogram` intervals? I did not see any rollup job configuration details to be 100% certain about what the issue is 🙂 

Querying rollup doc date histogram requires that the queried interval be a *multiple* of the configured rollup job interval. The queried interval must also be of the same *type* as the configured interval, see
this doc section about [Calendar vs Fixed time intervals](https://www.elastic.co/guide/en/elasticsearch/reference/master/rollup-job-config.html#_date_histogram_2).

For visualizations, we implemented the idea of a `baseInterval` that is passed to the UI input and becomes its default value. The `baseInterval` is the date histogram interval configured on the rollup job. If the user changes the value, then we check to see if their custom interval is a valid multiple of `baseInterval`:
https://github.com/elastic/kibana/blob/master/src/ui/public/validate_date_interval.js#L45
https://github.com/elastic/kibana/blob/master/src/ui/public/vis/lib/least_common_interval.ts#L24

If it is valid multiple, the query proceeds normally. If it is not a valid multiple, the input box gets into an error state and prevents the query from being submitted until the input is corrected.

alexwizp commented:
@jen-huang Please review the following file. The main idea here that if interval is not valid, we round the 'user' interval to valid interval for the rollup job. In means that for all cases we are working only with multiple intervals and for me this issue looks like a issue on the ES side. 
https://github.com/elastic/kibana/pull/28762/files#diff-d96441362ad5272ea37f2cc1ad91756f

@AlonaNadler I think we can easily check is it a TSVB problem or not.  All what you need to do it's just a create simple query for rollup data in Dev Tools -> Console and check is your query return data for that interval or not.

alexwizp committed:
Merge remote-tracking branch 'upstream/master' into rollup

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12224/)

sulemanof commented:
Hi @AlonaNadler ! 
First of all, here is a config for the rollup job:
![rollup_job_config](https://user-images.githubusercontent.com/31325372/52567475-78600a00-2e1d-11e9-895b-8c02912512b2.PNG)

We found out next inconsistency in the rollup_search endpoint:
1. the rollup_searc endpoint fails trying to set field to `value_count`:

![value_count_fails](https://user-images.githubusercontent.com/31325372/52567523-99c0f600-2e1d-11e9-93de-67ccbfa1418c.PNG)

while other fields, like `sum`, `avg`, `max`, `min` work fine with the same configuration:

![sum_success](https://user-images.githubusercontent.com/31325372/52567614-d1c83900-2e1d-11e9-937c-47563a9d9c98.PNG)

2. `rollup_search` endpoint skips pipeline aggregation (derivatives, cumulative sum) for a rollup index:

![rollup_pipeline_skip_aggjpg](https://user-images.githubusercontent.com/31325372/52567984-9a0dc100-2e1e-11e9-925f-93d17c2422e3.jpg)

though it works for basic index:

![search_pipeline_success](https://user-images.githubusercontent.com/31325372/52568032-be699d80-2e1e-11e9-8c1a-4396277ea2c3.jpg)


Please, provide next steps for us. Maybe it would be a good idea to include someone responsible for the `rollup_search` endpoint to provide details

alexwizp commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12264/)

cjcenizal commented:
@sulemanof FYI, Zach Tong on the ES team will be the person to speak to regarding the `_rollup_search` endpoint.

polyfractal commented:
Re: rollup search issues:

> 1. The rollup_search endpoint fails trying to set field to `value_count`:

`value_count` looks like a bug, we don't currently have it white-listed in the translation routine.  Should be easy to fix, I'll open a PR in a little bit.

> 2. `rollup_search` endpoint skips pipeline aggregation (derivatives, cumulative sum) for a rollup index:

Pipeline aggs missing is also a bug (documented here: https://github.com/elastic/elasticsearch/issues/34390, and related SDHE: https://github.com/elastic/support-dev-help-elasticsearch/issues/1011#issuecomment-428480674).  I have a WIP branch that adds support for pipelines, but it was really hacky and I never got a chance to finish it... or decide if we wanted to support such a hacky method.  

In either case, adding support for pipelines is going to take some work.  I believe TSVB makes heavy use of pipelines?  Would that be a blocker for TSVB-Rollup support?

AlonaNadler commented:
Thanks for the quick response @polyfractal 

> In either case, adding support for pipelines is going to take some work. I believe TSVB makes heavy use of pipelines? Would that be a blocker for TSVB-Rollup support?

Indeed TSVB makes heavy use of pipeline, there are several reasons why TSVB tend to be  popular in the metric use case:
* Recent values - metrics usually if more focused on what recently happened which TSVB shows in the metric, gauge, and top N
* Ability to calculate more advanced calculation which usually relies on pipeline aggregations
* TSVB supports more aggregations types - the more common is `filter ratio`, `bucket aggregation`. Filter ratio probably won't work since depended on grouping by a filter

I guess we can still benefit from the first item

sulemanof committed:
Merge remote-tracking branch 'upstream/master' into rollup

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12521/)

timroes commented on src/legacy/core_plugins/metrics/server/lib/vis_data/request_processors/table/normalize_query.js:
```diff
@@ -0,0 +1,48 @@
+
+const isEmptyFilter = (filter = {}) => Boolean(filter.match_all) && isEmpty(filter.match_all);
+
+export default function normalizeQuery() {
```
Same here, could we please add some documentation about the function (most likely a copy of the other file).

alexwizp committed:
remove todo

alexwizp committed:
Merge remote-tracking branch 'upstream/master' into rollup

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/883/)

alexwizp committed:
fix calculation of interval value for rollup search

alexwizp commented on src/legacy/core_plugins/metrics/server/lib/vis_data/get_series_data.js:
```diff
@@ -16,50 +16,44 @@
-        resp.type = panel.type;
+  const bodiesPromises = panel.series.map(series => getRequestParams(req, panel, series, esQueryConfig, capabilities));
+  const body = (await Promise.all(bodiesPromises))
+    .reduce((acc, items) => acc.concat(items), []);
```
.flat is not supported now

alexwizp commented on src/legacy/core_plugins/metrics/server/lib/vis_data/get_annotations.js:
```diff
@@ -1,84 +1,54 @@
+
+  const bodiesPromises = annotations.map(annotation => getRequestParams(req, panel, annotation, esQueryConfig, capabilities));
+  const body = (await Promise.all(bodiesPromises))
+    .reduce((acc, items) => acc.concat(items), []);
```
.flat is not supported now

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1119/)

alexwizp committed:
add unit tests

alexwizp commented on src/legacy/core_plugins/metrics/server/lib/vis_data/request_processors/series/normalize_query.js:
```diff
@@ -0,0 +1,39 @@
+  return doc;
+}
+
+export default function normalizeQuery(req, panel, series) {
```
Done

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1157/)

markov00 commented on x-pack/plugins/rollup/server/lib/search_strategies/rollup_search_strategy.js:
```diff
@@ -0,0 +1,67 @@
+
+      return callWithRequest(ROLLUP_INDEX_CAPABILITIES_METHOD, {
+        indexPattern,
+      }).catch(() => Promise.resolve({}));
```
Are we silently catching any error on any rollup call? does we have any notification of a wrong call/es status etc somewhere else?

alexwizp replied:
It's ok, we just resolve our promise with {} if we cannot find rollup job by passed indexPattern. In this case we use default search strategy.

markov00 reviewed (changes requested):
Code LGTM beside:
- a wide use of default exports thats needs to be substituted by named exports as in our [styleguide](https://github.com/elastic/kibana/blob/master/style_guides/js_style_guide.md#use-named-exports-only)
- some minor comments on `SearchRequest`, `parseInterval`, some argument deconstruction and abstract classes

I didn't run the code btw. Will do and post an updated review when done.

alexwizp commented on src/legacy/core_plugins/metrics/server/lib/vis_data/helpers/unit_to_seconds.js:
```diff
@@ -16,18 +16,41 @@
+
+export const convertIntervalToUnit = (intervalString, unit) => {
+  const parsedInterval = parseInterval(intervalString);
+  const value = Number((parsedInterval.value * units[parsedInterval.unit] / units[unit]).toFixed(2));
```
parseInterval will not return null
![image](https://user-images.githubusercontent.com/20072247/52951094-e95d6f80-3391-11e9-912c-338d32343bbf.png)

alexwizp commented on src/legacy/core_plugins/metrics/server/lib/search_strategies/searh_requests/search_request.js:
```diff
@@ -0,0 +1,37 @@
+
+export default class SearchRequest extends AbstractSearchRequest {
+  getSearchRequestType(options) {
+    const isMultiSearch = Array.isArray(options.body);
```
No, for single search body is object. See for more details
https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-body.html 

Please see get_table_data.js file. We use single search here

markov00 replied:
I mean: if we know that `options.body` is an array but with only one element we could extract that body from the array and send it as a single search.

sulemanof committed:
Remove default exports

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1196/)

sulemanof commented on src/legacy/core_plugins/metrics/server/lib/vis_data/annorations/build_request_body.js:
```diff
@@ -16,12 +16,11 @@
-
-export default function buildAnnotationRequest(req, panel, annotation,  esQueryConfig, indexPattern) {
-  const processor = buildProcessorFunction(processors, req, panel, annotation,  esQueryConfig, indexPattern);
+export default function buildAnnotationRequest(...args) {
```
I do not see a necessity to add jsdoc here, since all arguments just pass through this function call to the next function in the same order. But anyway jsdoc was added

alexwizp committed:
Merge remote-tracking branch 'upstream/master' into rollup

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1260/)

alexwizp committed:
fix PR comments

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1268/)

alexwizp commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1272/)

alexwizp committed:
fix calendar intervals

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1316/)

timroes reviewed (approved):
Code LGTM, In case @AlonaNadler is fine with the functionality that way, I am fine with merging this.

alexwizp committed:
Merge remote-tracking branch 'upstream/master' into rollup

# Conflicts:
#	src/legacy/core_plugins/metrics/server/lib/vis_data/get_annotations.js
#	src/legacy/core_plugins/metrics/server/lib/vis_data/get_table_data.js
#	src/legacy/core_plugins/metrics/server/lib/vis_data/series/get_request_params.js

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1563/)

alexwizp commented:
@cchaos could you please have a look? We need review from elastic/kibana-design

cchaos reviewed (approved):
Glanced at the SASS changes, just an addition of another selector. LGTM

AlonaNadler commented:
I checked the PR again, looks good
Few comments
* After selecting aggregation when selecting the field it shows me  each field with every metric  (e.g. `bytes.max_value` ), I understand it is the way the field appears in the document
![image](https://user-images.githubusercontent.com/32550316/53344687-2365ce80-38c8-11e9-9c68-f5276228f700.png)
In the classic visual editor, it doesn't appear in the same way, it only shows the field `bytes`, assuming users already selected aggregation can we do it similarly in TSVB 
![image](https://user-images.githubusercontent.com/32550316/53344872-89eaec80-38c8-11e9-9abe-0b6ff5daeef8.png)
* Can we have an error when selecting an aggregation that can't work? currently, it shows zero which might be misleading without any additional context or error
* I configured the index pattern in panel option to be on both rollup and regular index `roll1*,kibana_sample_data_logs` and specific series only on `rollup1*` minimum intervals are 30s.
- Does it make sense that when my minimal intervals are configured on 30s and  I'm zooming in to specific time range where TSVB selects 5s as intervals and I still see results on the series that uses only rollup?
![image](https://user-images.githubusercontent.com/32550316/53348073-3af48580-38cf-11e9-9cb5-22982a263673.png)

timroes commented:
> * After selecting aggregation when selecting the field it shows me  each field with every metric  (e.g. `bytes.max_value` ), I understand it is the way the field appears in the document

This should not happen. It should show only the "original field name", because that's the field name we need to query against. So it should behave like the classical vis editor as you've pointed out.

alexwizp commented:
> * After selecting aggregation when selecting the field it shows me  each field with every metric  (e.g. `bytes.max_value` ), I understand it is the way the field appears in the document

> This should not happen. It should show only the "original field name", because that's the field name we need to query against. So it should behave like the classical vis editor as you've pointed out.'

@timroes  @AlonaNadler  Let me explain. In example above Alona use the index pattern with wildcard (roll1*,kibana_sample_data_logs). From our requirements we cannot use Rollup search strategy for indexes with wildcard.  See selected text below:
![image](https://user-images.githubusercontent.com/20072247/53367361-04fae400-3957-11e9-839a-96f931945399.png)

For Rollup Search Strategy we reuse logic from classical vis editor.  See x-pack/plugins/rollup/server/lib/merge_capabilities_with_fields.js file in this PR. It means that if we detect that we should use Rollup Search Strategy we show `bytes`  and filter all extra fields like e.g. `bytes.max_value`. 

For Default Search Strategy all works like in the master branch. 
@AlonaNadler  @timroes  Could you please checkout the master branch (without my changes) and try to create the same visualization with (roll1*,kibana_sample_data_logs).  `bytes.max_value`  will be there

AlonaNadler commented:
Initially, the team who implemented rollups in Kibana had a requirement to not support wildcard in rollup index pattern in Kibana. Apparently, there was a change and it is  possible to use a wildcard with index pattern, 
I followed up with the team and they said :

> "we revised it so that it is possible to create a rollup index pattern with a wildcard now. the original reason was to prevent users from matching against more than one rollup index (since rollup search only supports querying against one rollup index). however, allowing users to use wildcard lets them search one rollup index + any number of regular indices. for example `rollup_logstash,logstash-*` as a single pattern"

I didn't' know about this was changed and in the rollup requirements, I added it as a requirement to not support wildcard in rollup, which was implemented in this PR, apologies my fault.

Is there any chance to revert this back?

alexwizp committed:
Merge branch 'master' into rollup

alexwizp commented:
It's not a problem but let me show you one case:
1) Open you fresh Kibana instance and install 2 of our sample data: e.g Flights and E-Commerce (two indexes should be created here: kibana_sample_data_flights, kibana_sample_data_ecommerce)
2) Go to TSVB and create new visualization with unsupported for Rollup aggregation (in my example I just set 'Filter' for the Group By field). Index pattern should be "kibana_*". 
3) Save you visualization. 

**Actual behavior:** user see data for all indexes with prefix "kibana_". In real world it can be e.g company name. TSVB works as expected. 
![image](https://user-images.githubusercontent.com/20072247/53397528-61dab680-39b8-11e9-94b2-9a7267d91748.png)


Let's go next: 

4) Open Management -> Rollup Job and create a Rollup index. Prefix should be like before "kibana_". For my example I created a rollup job with name "kibana_rollup"

5) Open visualization which we saved in step 3

**Expected behaviour:** I expect to see the aggregated data by kibana_* wildcard like in step 3
**Actual behaviour:** I see the Error Message that Filter aggregation is not supported by Rollup search. 
![image](https://user-images.githubusercontent.com/20072247/53397458-4374bb00-39b8-11e9-869b-de6b45a0acd2.png)

@AlonaNadler  @timroes  Are you ok with that? If not probably we don't need to revert it

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2479/)

rayafratkina commented:
We should have a separate conversation and PR about wildcard support. It's not worth holding up this PR while we figure out what do.

elasticmachine commented:
Pinging @elastic/kibana-app
